src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs

```diff
@@ -0,0 +1,114 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+using System.Timers;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.AutoML
+{
+    public interface IPerformanceMonitor : IDisposable
+    {
+        void Start();
+
+        void Stop();
+
+        double? GetPeakMemoryUsageInMegaByte();
+
+        double? GetPeakCpuUsage();
+
+        public event EventHandler<double> CpuUsage;
+
+        public event EventHandler<double> MemoryUsageInMegaByte;
+    }
+
+    internal class DefaultPerformanceMonitor : IPerformanceMonitor
+    {
+        private readonly IChannel _logger;
+        private Timer _timer;
+        private double? _peakCpuUsage;
+        private double? _peakMemoryUsage;
+        private readonly int _checkIntervalInMilliseconds;
+        private TimeSpan _totalCpuProcessorTime;
+
+        public DefaultPerformanceMonitor(IChannel logger, int checkIntervalInMilliseconds)
+        {
+            _logger = logger;
+            _checkIntervalInMilliseconds = checkIntervalInMilliseconds;
+        }
+
+
+        public event EventHandler<double> CpuUsage;
+
+        public event EventHandler<double> MemoryUsageInMegaByte;
+
+
+        public void Dispose()
+        {
+            Stop();
+        }
+
+        public double? GetPeakCpuUsage()
+        {
+            return _peakCpuUsage;
+        }
+
+        public double? GetPeakMemoryUsageInMegaByte()
+        {
+            return _peakMemoryUsage;
+        }
+
+        public void Start()
+        {
+            if (_timer == null)
+            {
+                _timer = new Timer(_checkIntervalInMilliseconds);
+                _totalCpuProcessorTime = Process.GetCurrentProcess().TotalProcessorTime;
+                _timer.Elapsed += OnCheckCpuAndMemoryUsage;
+                _timer.AutoReset = true;
+                _timer.Enabled = true;
+                _logger?.Trace($"{typeof(DefaultPerformanceMonitor)} has been started");
+            }
+        }
+
+        public void Stop()
+        {
+            _timer?.Stop();
+            _timer?.Dispose();
+            _timer = null;
+            _peakCpuUsage = null;
+            _peakMemoryUsage = null;
+        }
+
+        private void OnCheckCpuAndMemoryUsage(object source, ElapsedEventArgs e)
+        {
+            SampleCpuAndMemoryUsage();
+        }
+
+        private void SampleCpuAndMemoryUsage()
+        {
+            // calculate CPU usage in %
+            using (var process = Process.GetCurrentProcess())
+            {
+                var currentCpuProcessorTime = Process.GetCurrentProcess().TotalProcessorTime;
+                var elapseCpuProcessorTime = currentCpuProcessorTime - _totalCpuProcessorTime;
+                var cpuUsedMs = elapseCpuProcessorTime.TotalMilliseconds;
+                var cpuUsageInTotal = cpuUsedMs / (Environment.ProcessorCount * _checkIntervalInMilliseconds);
```

Comment:
>_checkIntervalInMilliseconds [](http://example.com/codeflow?start=80&length=28)

why multiply by _checkIntervalInMilliseconds?

Reply:
the % of cpu is calcuated by $(CPU time used by current process) / $(total CPU time)

for every `_checkIntervalInMilliseconds`
process uses `TotalProcessorTime - (last total processor time)`
total CPU time is # of processor count * `_checkIntervalInMilliseconds`

Reply:
ok. thanks!

Reply:
comment will be good here.

Reply:
### Resolved